File: ViewCs/Commands/SimpleCommand.cs
//  -*-  coding: utf-8-with-signature;  mode: c++  -*-  //
/*************************************************************************
**                                                                      **
**                  ---   Graphics Test Project.   ---                  **
**                                                                      **
**          Copyright (C), 2025-2026, Takahiro Itou                     **
**          All Rights Reserved.                                        **
**                                                                      **
**          License: (See COPYING or LICENSE files)                     **
**          GNU Affero General Public License (AGPL) version 3,         **
**          or (at your option) any later version.                      **
**                                                                      **
*************************************************************************/

using System;
using System.ComponentModel;
using System.Windows.Input;


namespace  ViewCs.Commands  {

//========================================================================
//
//    SimpleCommand  class.
//

public  class  SimpleCommand<T> : ICommand
{

//========================================================================
//
//    Constructor(s) and Destructor.
//

//----------------------------------------------------------------
/**   コンストラクタ。
**
**/
public SimpleCommand(
        Action<T>           execute,
        Predicate<object?>? canExecute = null)
{
    this.m_execute  = execute ?? throw new ArgumentNullException(
            nameof(execute));
    this.m_canExecute = canExecute;
}


//========================================================================
//
//    Public Member Functions (Implement Interface).
//

//----------------------------------------------------------------
/**   コマンドが実行可能か否かを返す。
**
**/
public  bool
CanExecute(object? parameter)
{
    return ( this.m_canExecute?.Invoke(parameter) ?? true );
}

//----------------------------------------------------------------
/**
**
**/
public  void
Execute(object? parameter)
{
    this.m_execute((T)parameter);
}


//========================================================================
//
//    Public Events (Implement Interface).
//

//----------------------------------------------------------------
/**
**
**/
public  event   EventHandler?   CanExecuteChanged;


//========================================================================
//
//    Public Member Functions.
//

//----------------------------------------------------------------
/**   CanExecuteChanged イベントを発生させる。
**
**/
public  void
RaiseCanExecuteChanged()
{
    CanExecuteChanged?.Invoke(this, EventArgs.Empty);
}


//========================================================================
//
//    Member Variables.
//

/**   実行する内容。    **/
private  readonly   Action<T>               m_execute;

/**   実行可否の判定。  **/
private  readonly   Predicate<object?>?     m_canExecute;

}   //  End class  SimpleCommand

}   //  End of namespace  ViewCs.Commands

File: ViewVb/Commands/SimpleCommand.vb
''  -*-  coding: utf-8-with-signature  -*-  ''
''************************************************************************
''                                                                      ''
''                  ---   Graphics Test Project.   ---                  ''
''                                                                      ''
''          Copyright (C), 2025-2026, Takahiro Itou                     ''
''          All Rights Reserved.                                        ''
''                                                                      ''
''          License: (See COPYING or LICENSE files)                     ''
''          GNU Affero General Public License (AGPL) version 3,         ''
''          or (at your option) any later version.                      ''
''                                                                      ''
''************************************************************************

Imports System
Imports System.ComponentModel
Imports System.Windows.Input


Namespace Global.ViewVb.Commands

''========================================================================
''
''    SimpleCommand  class.
''

Public Class SimpleCommand(Of T)
        Implements ICommand

''========================================================================
''
''    Member Variables.
''

Private  ReadOnly  m_execute    As Action(Of T)
Private  ReadOnly  m_canExecute As Predicate(Of Object)

Private  Shared    s_typeConverter As TypeConverter  _
    = TypeDescriptor.GetConverter(GetType(T))


''========================================================================
''
''    Constructor(s) and Destructor.
''

Public Sub New(
        execute As Action(Of T),
        Optional canExecute As Predicate(Of Object) = Nothing)
''--------------------------------------------------------------------
''    コンストラクタ
''--------------------------------------------------------------------
    Me.m_execute    = execute
    Me.m_canExecute = canExecute
End Sub


''========================================================================
''
''    Public Member Functions (Implement Interface).
''

Public Function CanExecute(parameter As Object) As Boolean _
        Implements ICommand.CanExecute
    Return If(m_canExecute Is Nothing, True, m_canExecute(parameter))
End Function


Public Sub Execute(parameter As Object) Implements ICommand.Execute
Dim tparam As T

    If TypeOf parameter is T Then
        tparam = CType(parameter, T)
    Else
        tparam = s_typeConverter.ConvertFrom(parameter)
    End If
    Me.m_execute(tparam)
End Sub


''========================================================================
''
''    Public Events (Implement Interface).
''

Public Event CanExecuteChanged As EventHandler _
        Implements ICommand.CanExecuteChanged


''========================================================================
''
''    Public Member Functions.
''

Public Sub raiseCanExecuteChanged()
    RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
End Sub


End Class

End Namespace
